File: mvn-advent-of-code/src/main/java/advent/twenty_fifteen/Day20InfiniteElvesAndInfiniteHouses.java
package advent.twenty_fifteen;

import advent.common.DailyProblem;

public class Day20InfiniteElvesAndInfiniteHouses implements DailyProblem<Integer, Integer> {
    private final int part1Answer;
    private final int part2Answer;

    public Day20InfiniteElvesAndInfiniteHouses(int target) {
        int MAX_HOUSE = 1000000;

        this.part1Answer = calculatePartOne(target, MAX_HOUSE);
        this.part2Answer = calculatePartTwo(target, MAX_HOUSE);
    }


    private static int calculatePartOne(int target, int houseCount) {
        int[] houses = new int[houseCount];
        for (int elf = 1; elf != houses.length; elf++) {
            int hIndex = elf;
            while (hIndex < houses.length) {
                houses[hIndex] += elf * 10;
                hIndex += elf;
            }
        }
        return findHouseIndex(target, houses);
    }

    private static int calculatePartTwo(int target, int houseCount) {
        int[] houses = new int[houseCount];
        int[] elfDrops = new int[houseCount];
        for (int elf = 1; elf != houses.length; elf++) {
            int hIndex = elf;
            while (hIndex < houses.length && elfDrops[elf] < 50) {
                houses[hIndex] += elf * 11;
                hIndex += elf;
                elfDrops[elf]++;
            }
        }
        return findHouseIndex(target, houses);
    }

    private static int findHouseIndex(int target, int[] houses) {
        for (int i = 1; i != houses.length; i++) {
            if (houses[i] > target) {
                return i;
            }
        }
        return -1;
    }

    @Override
    public Integer getPart1Answer() {
        return part1Answer;
    }

    @Override
    public Integer getPart2Answer() {
        return part2Answer;
    }
}

File: scala-advent-of-code/src/main/scala/advent/twenty_fifteen/Day20InfiniteElvesAndInfiniteHouses.scala
package advent.twenty_fifteen

import advent.common.DailyProblem
import advent.utilities.FileUtilities

class Day20InfiniteElvesAndInfiniteHouses(targetPresents : Int) extends DailyProblem[Int, Int] {

  def elfVisitPattern1(elf : Int, houseCount : Int) : List[Int] = {
    (elf to houseCount by elf).toList
  }

  def elfVisitPattern2(elf : Int, houseCount : Int) : List[Int] = {
    (elf to houseCount by elf).toList.take(50)
  }

  def populateHouse(houseCount : Int, pattern : (Int,Int) => List[Int], gifts: (Int) => Int) : Int = {
    val houses = new Array[Int](houseCount+1)
    var result = Integer.MAX_VALUE
    for (elf <- 1 to houseCount) {
      for (hIndex <- pattern(elf , houseCount)) {
        houses(hIndex) += gifts(elf)
        if (houses(hIndex) >= targetPresents)
          result = Math.min(hIndex, result)
      }
    }

    result
  }

  override val part1Answer: Int = populateHouse(1000000, elfVisitPattern1, e => e * 10)
  override val part2Answer: Int = populateHouse(1000000, elfVisitPattern2, e => e * 11)
}
